File: c++/fractionToDecimal.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <cmath>
#include <sstream>
#include <cstring>
#include <unordered_map>

using namespace std;

// change stringstream method to use to_string() function
// don't know why, fail this case, rethink about it tomorrow;
string fractionToDecimal(int numerator, int denominator) { 
    string result;
    if (denominator == 0) return "";
    if ( (numerator > 0 && denominator < 0) || (numerator < 0 && denominator > 0) )
        result += '-'; 
    long num = abs((long)numerator);
    long den = abs((long)denominator);
    result += to_string(num / den);
    if (num % den == 0) return result;

    result += '.'; 
    unordered_map<int, int> hashmap;
    long remainder;
    for(remainder = num % den; remainder; remainder %= den) {
        if (hashmap.count(remainder) > 0) { // if recurring, add ( at the first appeared place
            result.insert(hashmap[remainder], 1, '(');
            result += ')';  
            break;
        }
        hashmap[remainder] = result.size();
        remainder *= 10;
        result += to_string(remainder / den);
    } 
    return result;
}

int main(){
    cout << fractionToDecimal(-1, -2147483648) << endl;
    return 0;
}
/*
Input:	-1, -2147483648
Output:	"0.0000000000000000000000000000001"
Expected:	"0.0000000004656612873077392578125"
*/

File: LeetCode_Java_WinterBreak2014/fractionToDecimal.java
import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;
//import java.lang.AbstractStringBuilder;

public class fractionToDecimal {
    public static class Solution {
        public String fractionToDecimal(int numerator, int denominator) {
            StringBuffer result = new StringBuffer();
            if (denominator == 0) {
                return "";
            }
            if ((numerator > 0 && denominator < 0)
                || (numerator < 0 && denominator > 0)) {
                result.append('-');
            }
            long num = Math.abs((long)numerator);
            long den = Math.abs((long)denominator);
            result.append((num / den));
            if (num % den == 0) {
                return result.toString();   
            }

            result.append('.');
            Map<Long, Integer> hash = new HashMap<Long, Integer>();
            long remainder;
            for (remainder = num % den; remainder > 0; remainder %= den) {
                if (hash.containsKey(remainder)) {
                    result.insert(hash.get(remainder), "(");  // insert overload function '('
                    result.append(')');
                    break;
                }
                hash.put(remainder, result.length());
                remainder *= 10;
                result.append(remainder / den);
            }
            return result.toString();
        }
    }

    public static void main(String[] args){
        Solution result = new Solution();
        String res = result.fractionToDecimal(2, 6);
        System.out.println(res);
    }
}
